dns_server/src/crawler/core/mod.rs

```diff
@@ -0,0 +1,582 @@
+
+/// Address state transition
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum AddressTransition {
+    Disconnected,
+    Connecting,
+    Connected,
+}
+
+/// Connection state of a potential node address (outbound only)
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+enum AddressState {
+    Disconnected {
+        /// The number of consecutive failed connection attempts.
+        /// New connection attempts are made after a progressive backoff time.
+        fail_count: u32,
+    },
+    Connecting {
+        /// Same as in the Disconnected state
+        fail_count: u32,
+    },
+    Connected {},
+}
+
+/// Additional state of a potential node address
+struct AddressData {
+    /// Connection state
+    state: AddressState,
+
+    /// Last time when the connection state was updated
+    state_updated_at: Duration,
+
+    /// Whether the address was reachable at least once.
+    /// Addresses that were once reachable are stored in the DB.
+    was_reachable: bool,
+
+    /// Whether the address was added from the command line
+    user_added: bool,
+}
+
+#[derive(Clone)]
+pub struct CrawlerCoreConfig {
+    pub p2p_port: u16,
+}
+
+pub struct CrawlerCore<N: NetworkingService, C> {
+    /// Time of some monotonic timer, started from 0
+    now: Duration,
+
+    /// Crawler config
+    config: CrawlerCoreConfig,
+
+    /// Chain config
+    chain_config: Arc<ChainConfig>,
+
+    /// Map of all known addresses (including currently unreachable)
+    addresses: BTreeMap<N::Address, AddressData>,
+
+    /// Map of all currently connected outbound peers
+    outbound_peers: BTreeMap<N::PeerId, N::Address>,
+
+    /// CoreCallback implementation
+    _phantom: PhantomData<C>,
+}
+
+pub enum CoreEvent<N: NetworkingService> {
+    Timer { period: Duration },
+    Connectivity { event: ConnectivityEvent<N> },
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum CoreCommand<N: NetworkingService> {
+    DnsAddAddress(IpAddr),
+    DnsDelAddress(IpAddr),
+
+    DbStoreAddress(N::Address),
+    DbEraseAddress(N::Address),
+
+    Connect(N::Address),
+    Disconnect(N::PeerId),
+    SendResponse(N::PeerId, PeerManagerMessage),
+}
+
+pub trait CoreCallback<N: NetworkingService> {
+    fn exec(&mut self, command: CoreCommand<N>);
+}
+
+impl<N: NetworkingService, C: CoreCallback<N>> CrawlerCore<N, C> {
+    pub fn new(
+        config: CrawlerCoreConfig,
+        chain_config: Arc<ChainConfig>,
+        loaded_addresses: BTreeSet<N::Address>,
+        added_addresses: BTreeSet<N::Address>,
+    ) -> Self {
+        let now = Duration::ZERO;
+
+        let addresses = loaded_addresses
+            .union(&added_addresses)
+            .map(|addr| {
+                (
+                    addr.clone(),
+                    AddressData {
+                        state: AddressState::Disconnected { fail_count: 0 },
+                        state_updated_at: now,
+                        was_reachable: loaded_addresses.contains(addr),
```

Comment:
This is such a weird way for initialization... why not initialize each alone, then chain them together instead of doing a union and finding out where they come from?

Reply:
It was done such way before, but user added addresses and addresses from the DB could be in both lists at the same time. And when you construct `BTreeMap` from such a chained iterator, some of the flags get lost (`was_reachable` or `user_added` depending on the chaining order). This caused problems in tests.

Reply:
I see.